Add a keycloak module to query keys metadata

thomasbach-dev opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

<!--- Please do not forget to include a changelog fragment:
      https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-changelog-fragments
      No need to include one for docs-only or test-only PR, and for new plugin/module PRs.
      Read about more details in CONTRIBUTING.md.
      -->

When setting up Nextcloud integration to Keycloak we have to ship the metadata of a realm to Nextcloud. With the current modules it is not possible to extract the metadata. This PR adds a module `keycloak.keycloak_realm_keys_metadata` which makes it possible to query the metadata and pass it on:

```yaml
- name: Fetch Keys metadata
  community.general.keycloak_realm_keys_metadata:
    auth_keycloak_url: https://auth.example.com/auth
    auth_realm: master
    auth_username: USERNAME
    auth_password: PASSWORD
    realm: MyCustomRealm
  delegate_to: localhost
  register: keys_metadata

- name: Write the Keycloak keys certificate into a file
  ansible.builtin.copy:
    dest: /tmp/keycloak.cert
    content: |
      {{ keys_metadata['keys_metadata']['keys']
      | selectattr('algorithm', 'equalto', 'RS256')
      | map(attribute='certificate')
      | first
      }}
  delegate_to: localhost
```

The module presented here is already used in production in our Ansible playbooks and is known to work.

##### ISSUE TYPE
<!--- Pick one or more below and delete the rest.
      'Test Pull Request' is for PRs that add/extend tests without code changes. -->
- New Module/Plugin Pull Request

##### COMPONENT NAME
<!--- Write the SHORT NAME of the module, plugin, task or feature below. -->

`keycloak_realm_keys_metadata`

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

thomasbach-dev commented:
I added myself to the list of maintainers in the KeyCloak team to get sanity checks to get green.
I am using Keycloak and Ansible heavily in a current project. So I actually might have more contributions. :)

russoz commented on plugins/module_utils/identity/keycloak/keycloak.py:
```diff
@@ -306,6 +307,32 @@ def get_realm_info_by_id(self, realm='master'):
+        """ Obtain realm public info by id
+
+        :param realm: realm id
+        :return: dict of real, representation or None if none matching exist
```
Typo?
```suggestion
        :return: dict with realm configuration, representation or None if none matching exist
```

thomasbach-dev replied:
Looks like a typo, right? :) tbh, I just copy pasted this from other descriptions and did not really get what `dict of real` could possibly mean. It makes sense now -- the typo propagated.
I'll have a proper look again what the returning data type is here

russoz replied:
Looks better. :-)

russoz commented on plugins/modules/keycloak_realm_keys_metadata.py:
```diff
@@ -0,0 +1,143 @@
+
+short_description: Allows obtaining Keycloak realm keys metadata via Keycloak API
+
+version_added: 9.1.0
```
9.1.0 has already been released, this module will go into the next one
```suggestion
version_added: 9.2.0
```

thomasbach-dev replied:
Addressed in a9de64c1937efd3db2ef7608485987f8b698adb2

russoz reviewed (commented):
Hi @thomasbach-dev 
Thanks for your contribution. There is a couple of adjustments to make, other than that it looks  good.

felixfontein commented on plugins/modules/keycloak_realm_keys_metadata.py:
```diff
@@ -0,0 +1,143 @@
+      be returned that way by this module. You may pass single values for attributes when calling the module,
+      and this will be translated into a list suitable for the API.
+
+attributes:
+    check_mode:
+        support: full
+    diff_mode:
+        support: full
+
```
For `_info` modules, remove this:
```suggestion
```

thomasbach-dev replied:
Adressed in a9de64c1937efd3db2ef7608485987f8b698adb2

felixfontein commented on plugins/modules/keycloak_realm_keys_metadata.py:
```diff
@@ -0,0 +1,143 @@
+
+extends_documentation_fragment:
+    - community.general.keycloak
+    - community.general.attributes
```
... and add this instead:
```suggestion
    - community.general.attributes
    - community.general.attributes.info_module
```

thomasbach-dev replied:
Addressed in a9de64c1937efd3db2ef7608485987f8b698adb2

felixfontein reviewed (commented):
Thanks for your contribution! Please note that every new module must come with tests (unit tests, integration tests, or both).

thomasbach-dev commented on plugins/modules/keycloak_realm_keys_metadata.py:
```diff
@@ -0,0 +1,143 @@
+
+DOCUMENTATION = """
+---
+module: keycloak_realm_keys_metadata
```
Addressed in a9de64c1937efd3db2ef7608485987f8b698adb2 and a87ba3e2968281d7a97a58efa91930d49561de9e

thomasbach-dev commented:
Sorry for the late reply -- very busy weeks. I incorporated the easy fixes for now. I will properly document the return type of the added method in `keycloak.py` and see how I can test this.

thomasbach-dev commented:
I added a test in 2019594c5d56650b742912cbf718519cb20e4660 following the testing pattern I found in other `test_keycloak_*` modules.

russoz commented on plugins/modules/keycloak_realm_keys_metadata_info.py:
```diff
@@ -0,0 +1,141 @@
+keys_metadata:
+    description:
+
+        - Representation of the realm keys metadata (see
+          https://www.keycloak.org/docs-api/latest/rest-api/index.html#KeysMetadataRepresentation).
```
```suggestion
        - Representation of the realm keys metadata (see
          U(https://www.keycloak.org/docs-api/latest/rest-api/index.html#KeysMetadataRepresentation)).
```

thomasbach-dev replied:
Picked up in c6bee82ea1235c29bdf27c10ca1447caa149925b

russoz commented on plugins/modules/keycloak_realm_keys_metadata_info.py:
```diff
@@ -0,0 +1,141 @@
+
+short_description: Allows obtaining Keycloak realm keys metadata via Keycloak API
+
+version_added: 9.2.0
```
9.2.0 has been released, moving to next one
```suggestion
version_added: 9.3.0
```

thomasbach-dev replied:
Picked up in c6bee82ea1235c29bdf27c10ca1447caa149925b

russoz commented on plugins/modules/keycloak_realm_keys_metadata_info.py:
```diff
@@ -0,0 +1,141 @@
+
+
+def main():
+    """
+    Module execution
+
+    :return:
+    """
```
Personally I think this comment is redundant. It does not add much, IMHO.

thomasbach-dev replied:
Yes, I agree. I removed the comment in c6bee82ea1235c29bdf27c10ca1447caa149925b

russoz commented on tests/unit/plugins/modules/test_keycloak_realm_keys_metadata_info.py:
```diff
@@ -0,0 +1,185 @@
+    """
+
+    obj = keycloak_realm_keys_metadata_info.KeycloakAPI
+    with patch.object(
+        obj, "get_realm_keys_metadata_by_id", side_effect=get_realm_keys_metadata_by_id
+    ) as mock_get_realm_keys_metadata_by_id:
```
I think the readability of this code would be improved if this `with``were to stay in only one line. Rename the mock object to something shorter - this symbol name is used only in the very next line, having an explicit name does not add more clarity than the line where it is defined, IMHO.

thomasbach-dev replied:
Well, I use black to automatically format my code when saving. I had to disable it in this case, as I would not get the line below 80 characters -- which is the default max line length in black. Unless I am missing something, there is no style guideline for these things in this repo, right?

c6bee82ea1235c29bdf27c10ca1447caa149925b contains a version which looks like this:
```python
@contextmanager
def patch_keycloak_api(side_effect):
    # …
    obj = keycloak_realm_keys_metadata_info.KeycloakAPI
    with patch.object(obj, "get_realm_keys_metadata_by_id", side_effect=side_effect) as obj_mocked:
        yield obj_mocked
```

where the long line is 99 characters wide which seems to be accepted style. :)

russoz reviewed (commented):
Hi @thomasbach-dev , thanks for the updates. There is a couple of other minor stuff in there, but overall looks good.

russoz reviewed (approved):
LGTM

felixfontein commented on plugins/modules/keycloak_realm_keys_metadata_info.py:
```diff
@@ -0,0 +1,136 @@
+    type: dict
+    contains:
+        active:
+            description: A mapping (i.e. a dict) from key algorithms to UUIDs.
```
https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#avoid-using-latin-phrases
```suggestion
            description: A mapping (that is, a dict) from key algorithms to UUIDs.
```

thomasbach-dev replied:
Incorporated in latest fixup commit a8370a3101f7b38fb5c48c64c4ab2e947e481ebe

felixfontein commented on plugins/modules/keycloak_realm_keys_metadata_info.py:
```diff
@@ -0,0 +1,136 @@
+            returned: always
+        keys:
+            description: A list of dicts providing detailed information on the keys.
+            type: list
```
`type=list` should always come with `elements`:
```suggestion
            type: list
            elements: dict
```

thomasbach-dev replied:
Incorporated in latest fixup commit a8370a3101f7b38fb5c48c64c4ab2e947e481ebe

felixfontein commented on plugins/modules/keycloak_realm_keys_metadata_info.py:
```diff
@@ -0,0 +1,136 @@
+    auth_password: PASSWORD
+    realm: MyCustomRealm
+  delegate_to: localhost
+  register: keys_metadata
```
How about adjusting this name a bit...
```suggestion
  register: keycloak_keys_metadata
```

thomasbach-dev replied:
Sounds good to me, also incorporated in a8370a3101f7b38fb5c48c64c4ab2e947e481ebe

felixfontein commented on plugins/modules/keycloak_realm_keys_metadata_info.py:
```diff
@@ -0,0 +1,136 @@
+  ansible.builtin.copy:
+    dest: /tmp/keycloak.cert
+    content: |
+      {{ keys_metadata['keys_metadata']['keys']
```
... so that here it's more clear which of the two `keys_metadata` is the return value of the module? Might be helpful for beginners.
```suggestion
      {{ keycloak_keys_metadata['keys_metadata']['keys']
```

thomasbach-dev replied:
a8370a3101f7b38fb5c48c64c4ab2e947e481ebe

felixfontein commented on plugins/modules/keycloak_realm_keys_metadata_info.py:
```diff
@@ -0,0 +1,136 @@
+
+    - Attributes are multi-valued in the Keycloak API. All attributes are lists of individual values and will
+      be returned that way by this module. You may pass single values for attributes when calling the module,
+      and this will be translated into a list suitable for the API.
```
Is this last paragraph useful for this module? I'm wondering whether we should remove it for the `_info` modules. On the other hand, it doesn't really hurt either...

thomasbach-dev replied:
I removed the paragraph. I think it does not belong here, if it's not relevant for this module. Thanks for pointing this out. Fixed in a8370a3101f7b38fb5c48c64c4ab2e947e481ebe

thomasbach-dev commented:
So, yet another round: I incorporated the suggestions from @felixfontein -- thanks for the thorough examination! I added these as a fixup commit which I will squash if I get a thumbs up. :)
I also rebased on `main`.

russoz commented:
Still LGTM

felixfontein commented:
@thomasbach-dev no need to squash, since we squash on merge anyway :) But if you prefer to squash before merge you can also do that...

thomasbach-dev commented:
Hmm, okay. So is this ready to get merged? Is there something left to do from my side?

felixfontein commented:
No, just a last quick look and then merge :)

@thomasbach-dev thanks for your contribution!
@russoz thanks for reviewing!

felixfontein commented:
FYI @thomasbach-dev you might be interested in https://forum.ansible.com/t/keycloak-modules-in-community-general/41746
